### ISSUE TITLE
Parallel Trainer on GPU slow for RNN classification 

### ISSUE THREAD

**Issue by lacic** [deeplearning4j/deeplearning4j#3217]
From Gitter chat:

Hi, could you please give me your opinion about my GPU setup and runtime performance for training RNNs to predict sequences? Sorry for the wall of text which is following now! :)

Data:

I have 20 different classes which are used in a time series example (e.g., A -> B -> D -> A -> G -> ....). So basically, a similar problem as the character prediction example in dl4j github repo.

There are 25,915 different time series examples of different length at my disposal. I have made a 60/40 split to be used to train in an early-stopping fashion. This results into 15,549 different time series examples to train on and 10.365 to use for validation.

Each features/labels csv file is already written in form of a one-hot-encoding. For example, train/features/0.csv would look like:

1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0
0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0
…..

Training:

Here is the gist of the [code](https://gist.github.com/lacic/3ada1ae9b5abbc400acc28411b4a3df0) to train the network. The latest dl4j is used and with nd4j-cuda-8.0-platform.

The training is done on a GPU server with 2 x NVIDIA Tesla P100.

Now, to my questions. :)

1) With such data, code and gpu environment, training one Epoch with EarlyStoppingParallelTrainer takes around 2.5 – 3.5 hours.

Is that an expected and actually OK runtime performance for 1 epoch?

2) Within the (give or take) 3 hours needed for 1 epoch, around 45-50 minutes is needed for validating the model at the end of the epoch training (i.e., report something like EarlyStoppingParallelTrainer - New best model: score = 70.90593583163113, epoch = 6 (previous: score = 74.41270532552284, epoch = 5)).

Is there anything (besides making the validation set smaller) to speed-up the validation time?

3) I noticed that (by running nvidia-smi) the GPU-Utilization is mostly < 8%. Also, from the 2 GPUs, one is almost always at 0% GPU Utilization. In contrast to this, both use the most part of the available memory:

GPU_1 = 13673MiB / 16276MiB
GPU_2 = 13179MiB / 16276MiB

I even get here and there errors like:
“[ParallelWrapper trainer 3] WARN o.n.j.handler.impl.CudaZeroHandler - Out of [DEVICE] memory, host memory will be used instead: deviceId: [1], requested bytes: [417960000]”

Should I maybe try different parameters for the CudaEnvironment cache and grid size, or the numbers for workers and prefetch buffer in the EarlyStoppingParallelTrainer? Change the batch size?

I would appreciate if you could give me any tips on how to proceed further with optimizing my gpu trainer :D

**Comment by lacic** [comment]
As by gitter chat: "this sounds really slow"  :/

**Comment by raver119** [comment]
Aha, so you''re getting pushed out of gpu vram.

> “[ParallelWrapper trainer 3] WARN o.n.j.handler.impl.CudaZeroHandler - Out of [DEVICE] memory, host memory will be used instead: deviceId: [1], requested bytes: [417960000]”

Do you use some CudaEnvironment while getting such messages?

**Comment by lacic** [comment]
When you look at the above linked [code](https://gist.github.com/lacic/3ada1ae9b5abbc400acc28411b4a3df0) you can see that I define the following CudaEnvironment:

`CudaEnvironment.getInstance().getConfiguration()`
`.allowMultiGPU(true)`
` .setMaximumDeviceCacheableLength(1024 * 1024 * 1024L)`
` .setMaximumDeviceCache(6L * 1024 * 1024 * 1024L)`
`.setMaximumHostCacheableLength(1024 * 1024 * 1024L)`
`   .setMaximumHostCache(6L * 1024 * 1024 * 1024L)`
` .setMaximumGridSize(512)`
` .setMaximumBlockSize(512)`         
`.allowCrossDeviceAccess(true);`

**Comment by lacic** [comment]
Would it help if I prepare for you the train/validation data to test it out using the linked gist code?

**Comment by agibsonccc** [comment]
@lacic that would be great!

**Comment by lacic** [comment]
Hi! 

Here is the dataset I used in the above code to train the network: [dl4j_issue_3217_data.tar.gz](https://github.com/deeplearning4j/deeplearning4j/files/909404/dl4j_issue_3217_data.tar.gz)

If you extract it you should have the data_split/train and data_split/validation folders with the respective features/labels folders and files.

**Comment by raver119** [comment]
Please change cache settings to:
```
.setMaximumDeviceCacheableLength(32 * 1024 * 1024L)
.setMaximumDeviceCache(1L * 1024 * 1024 * 1024L)
```

and run once again.

**Comment by lacic** [comment]
Hi,

just to give you an update:

I've started the training again with the changed settings and java -Xms30g -Xmx30g -XX:+UseG1GC  flags. 

GPU Memory is 5791MiB / 16276MiB (GPU 1) and 6831MiB / 16276MiB (GPU 2).

The GPU utilization on every gpu is mostly between 0% and 4% (there were some spikes initially on one card between 13% and 24%).

The current output is:

`11:42:26.445 [main] DEBUG o.n.j.c.CudaAffinityManager - Manually mapping thread [51] to device [0], out of [2] devices...`
`11:42:49.534 [ParallelWrapper trainer 0] INFO  org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 0`
`11:48:37.433 [ParallelWrapper trainer 0] INFO  o.d.o.l.ScoreIterationListener - Score at iteration 0 is 612.2542128946592`
`12:10:29.517 [main] INFO  o.d.parallelism.ParallelWrapper - Averaged score: 624.5984431057364`
`12:48:09.454 [main] INFO  o.d.parallelism.ParallelWrapper - Averaged score: 715.9200443959036`
`13:16:17.781 [main] INFO  o.d.parallelism.ParallelWrapper - Averaged score: 666.1897977612091`
`13:45:43.834 [main] INFO  o.d.parallelism.ParallelWrapper - Averaged score: 642.573374608268`

No "Out of [DEVICE] memory" up until now, but running one epoch is still taking long (almost 2 hours now for the 1st epoch) with the given dataset.

**Comment by raver119** [comment]
Great, thanks. I'll investigate your issue once workspaces are merged.

**Comment by lacic** [comment]
Update 2:

Scoring of the 1st epoch has finished:

`13:55:32.790 [main] DEBUG o.d.parallelism.ParallelWrapper - Iterations passed: 13`
`13:55:32.790 [main] INFO  o.d.p.EarlyStoppingParallelTrainer - Completed training epoch 0`
`14:39:50.419 [main] INFO  o.d.p.EarlyStoppingParallelTrainer - Score at epoch 0: 581.5361699802916`
`14:39:52.180 [main] DEBUG o.n.j.c.CudaAffinityManager - Manually mapping thread [558] to device [0], out of [2] devices...`
`14:55:27.777 [main] INFO  o.d.parallelism.ParallelWrapper - Averaged score: 433.20138702997264`
`15:27:00.113 [main] INFO  o.d.parallelism.ParallelWrapper - Averaged score: 382.72005068321`
`15:59:59.300 [main] INFO  o.d.parallelism.ParallelWrapper - Averaged score: 277.80238526029973`

If we disregard the training time, the time it takes from completing a training epoch to scoring it is on its own really long. In this case it took from 13:55 until 14:39, that is 44 min? O_o

**Comment by raver119** [comment]
> Manually mapping thread [558] to device [0], out of [2] devices...

558 threads? What actually happens there? You just use ParallelWrapper and nothing else?

**Comment by lacic** [comment]
It's exactly the code linked in the gist above. 

If it helps, I could make a copy of the whole maven project to clone/run?

**Comment by lacic** [comment]
@raver119 Is that for certain that based on the output 558 threads are generated?

I've just tried to run it on the laptop (GeForce 940MX DDR5) with a really small sub-dataset. The output in the eclipse console and number of threads reported by jconsole do not match.  

Here is the [screen shot](http://i.imgur.com/xPrMxg5.png)

**Comment by raver119** [comment]
I think after workspaces + cuDNN lstm + gradients sharing introduction - this issue isn't relevant anymore.
